fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerMultiTargetTest.java

```diff
@@ -775,4 +777,153 @@ public void benchMarkMultiTarget() {
+            assertTrue(e.getMessage().contains("This index was partly built, and blocked"));
+        }
+
+        // 5. continue with unblock
+        try (OnlineIndexer indexBuilder = newIndexerBuilder()
+                .setTargetIndexes(indexes)
+                .setTimer(timer)
+                .setLimit(chunkSize)
+                .setIndexingPolicy(OnlineIndexer.IndexingPolicy.newBuilder()
+                        .setAllowUnblock(true)
+                        .build())
+                .build()) {
+
+            indexBuilder.buildIndex();
+        }
+
+        // validate
+        validateIndexes(indexes);
+    }
+
+    @Test
+    void testMultiTargetIndexingBlockerExpiration() {
+        // block indexing with a short ttl, continue after sleep
+        final FDBStoreTimer timer = new FDBStoreTimer();
+        final int numRecords = 59;
+        final int chunkSize  = 13;
+
+        List<Index> indexes = new ArrayList<>();
+        indexes.add(new Index("indexD", new GroupingKeyExpression(EmptyKeyExpression.EMPTY, 0), IndexTypes.COUNT));
+        indexes.add(new Index("indexA", field("num_value_2"), EmptyKeyExpression.EMPTY, IndexTypes.VALUE, IndexOptions.UNIQUE_OPTIONS));
+        indexes.add(new Index("indexB", field("num_value_3_indexed"), IndexTypes.VALUE));
+        indexes.add(new Index("indexC", field("num_value_unique"), EmptyKeyExpression.EMPTY, IndexTypes.VALUE, IndexOptions.UNIQUE_OPTIONS));
+
+        openSimpleMetaData();
+        populateData(numRecords);
+
+        FDBRecordStoreTestBase.RecordMetaDataHook hook = allIndexesHook(indexes);
+        openSimpleMetaData(hook);
+        disableAll(indexes);
+
+        // 1. partly build multi
+        buildIndexAndCrashHalfway(chunkSize, 2, timer, newIndexerBuilder()
+                .setTargetIndexes(indexes));
+
+        // 2. Block indexing
+        FDBRecordStoreTestBase.RecordMetaDataHook localHook = allIndexesHook(indexes);
+        openSimpleMetaData(localHook);
+        String luka = "my name is Luka";
+        try (OnlineIndexer indexer = OnlineIndexer.newBuilder()
+                .setDatabase(fdb).setMetaData(metaData).setSubspace(subspace)
+                .setTargetIndexes(indexes)
+                .build()) {
+            indexer.indexingStamp(OnlineIndexer.IndexingStampOperation.BLOCK, luka, 2);
+        }
+
+        // 3. fail continue without unblock
+        try (OnlineIndexer indexBuilder = newIndexerBuilder()
+                .setTargetIndexes(indexes)
+                .setTimer(timer)
+                .setLimit(chunkSize)
+                .build()) {
+            RecordCoreException e = assertThrows(RecordCoreException.class, indexBuilder::buildIndex);
+            assertTrue(e.getMessage().contains("This index was partly built, and blocked"));
+        }
+
+        // 4. sleep
+        try {
+            Thread.sleep(2000);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
```

Comment:
Does the record layer have any mechanisms for mocking out time? If we did, we wouldn't need to put a 2 second sleep in this test.

Otherwise, using millis for the expiration time would allow an expiration of 1 millisecond which wouldn't require so much sleeping here.

Reply:
What's the time mocking mechanism?